src/Group.jl

```diff
@@ -0,0 +1,397 @@
+"""
+    AbstractGroupOperation
+
+Abstract type for smooth binary operations on elements of a Lie group.
+"""
+abstract type AbstractGroupOperation end
+
+@doc doc"""
+    AbstractGroupManifold{<:AbstractGroupOperation} <: Manifold
+
+Abstract type for a Lie group (a group that is also a smooth manifold with a
+smooth binary operation). `AbstractGroupManifold`s must implement at least
+`inv`(@ref), `identity`(@ref), and `left_action`(@ref).
+
+Group manifolds by default assume a [`LeftInvariantCanonicalMetric`](@ref)
+$g_x(v,w) = g_e(x \cdot v, x \cdot w) = (x \cdot v)^T (x \cdot w)$, where
+$g_x$ is the metric tensor at a point $x$, $g_e$ is the metric tensor at the
+identity element $e$, and $x \cdot v$ is the left action of the $x$ on the
+tangent vector $v$. This behavior can be changed by reimplementing
+[`inner`](@ref), [`exp!`](@ref), and [`log!`](@ref).
+"""
+abstract type AbstractGroupManifold{O<:AbstractGroupOperation} <: Manifold end
+
+"""
+    GroupManifold{M<:Manifold,O<:AbstractGroupOperation} <: AbstractGroupManifold{O}
+
+Decorator for a smooth manifold that equips the manifold with a group
+operation, thus making it a Lie group. See [`AbstractGroupManifold`](@ref) for
+more details.
+
+# Constructor
+
+    GroupManifold(manifold, op)
+"""
+struct GroupManifold{M<:Manifold,O} <: AbstractGroupManifold{O}
+    manifold::M
+    op::O
+end
+
+@traitimpl IsDecoratorManifold{GroupManifold}
+
+struct LeftInvariantCanonicalMetric <: RiemannianMetric end
+struct RightInvariantCanonicalMetric <: RiemannianMetric end
+
+@traitimpl HasMetric{GroupManifold,LeftInvariantCanonicalMetric}
+
+"""
+    IsMatrixGroup
+
+A `Trait` to indicate that a group's default representation is a matrix
+representation. For groups with a `MultiplicationOperation`, this indicates
+that the group is a subgroup of the General Linear Group.
+"""
+@traitdef IsMatrixGroup{G}
+
+@doc doc"""
+    inv(G::AbstractGroupManifold, x)
+
+Inverse $x^{-1}$ of an element $x$, such that
+$x \cdot x^{-1} = x^{-1} \cdot x = e$.
+"""
+function inv(G::AbstractGroupManifold, x)
+    error("inv not implemented on $(typeof(G)) for point $(typeof(x))")
+end
+
+@doc doc"""
+    identity(G::AbstractGroupManifold, x)
+
+Identity element $e$, such that for any element $x$,
+$x \cdot e = e \cdot x = x$. The returned element is of a similar type to `x`.
+"""
+function identity(G::AbstractGroupManifold, x)
+    error("identity not implemented on $(typeof(G)) for point $(typeof(x))")
+end
+
+@doc doc"""
+    left_action(G::AbstractGroupManifold, x, p)
+
+For an element $p$ of some set, compute the left action of group element $x$
+on $p$, i.e. $x \cdot p$.
+"""
+function left_action(G::AbstractGroupManifold, x, p)
```

Comment:
It would be nice to have a motivating example for this. For all the Lie groups and G-spaces I've used, the action on the set is the same as the action upon itself.

Reply:
Right, it wasn't clear from the docstring that you only consider actions on `G`. In general, you could consider for example actions of SO(2) on R^3 and then you need to additionally specify the axis of rotation somewhere. I have a similar type in development for FunManifolds and it's actually useful.

Reply:
Ah, I see what you're saying. Yes, that would be useful.